File: ab-cloud-verification/haskell/VerifyRU.hs
module VerifyRU
  ( objection1RU
  , objection2RU
  , objection3RU
  ) where

import Data.List (sort)
import ZerosLoader (gramPoint, gueCDF)

-- | Возражение 1: сходимость b(N) (русский).
objection1RU :: [Double] -> String
objection1RU [] = "Нули не загружены.\n"
objection1RU zeros = unlines
  [ "╔══════════════════════════════════════════════════════╗"
  , "║  ВОЗРАЖЕНИЕ 1: Проверка сходимости b(N)             ║"
  , "╚══════════════════════════════════════════════════════╝"
  , ""
  , "  b(N) = (1/N) × Σ|γ_k - γ̃_k|"
  , "  Точки Грама γ̃_k: W Ламберта (Холли) + ньютоновская доводка"
  , ""
  , "  Загружено нулей: " ++ show n
  , ""
  , "  Таблица сходимости:"
  , "  ─────────────────────────────────"
  , "       N           b(N)"
  , "  ─────────────────────────────────"
  ] ++ tableLines ++ [
    "  ─────────────────────────────────"
  , ""
  , verdict
  , ""
  , "  Отклонение от точек Грама систематически убывает,"
  , "  подтверждая согласие нулей с законом Грама."
  ]
  where
    n = length zeros
    checks = filter (<= n) [100, 500, 1000, 5000, 10000, 50000, 100000, 500000, 1000000]
    bN cp = sum [abs (zeros !! k - gramPoint k) | k <- [0..cp-1]] / fromIntegral cp
    tableLines = [fmtRow cp (bN cp) | cp <- checks]
    totalBN = sum [abs (zeros !! k - gramPoint k) | k <- [0..n-1]] / fromIntegral n
    verdict = if totalBN < 0.5
      then "  ✓ b(N) = " ++ f6 totalBN ++ " → СХОДИМОСТЬ ПОДТВЕРЖДЕНА (b(N) → 0)"
      else "  ✗ b(N) = " ++ f6 totalBN ++ " → сходимость не выявлена"

-- | Возражение 2: KS-критерий GUE (русский).
objection2RU :: [Double] -> String
objection2RU zs
  | length zs < 2 = "Требуется ≥ 2 нулей для KS-критерия GUE.\n"
objection2RU zeros = unlines
  [ "╔══════════════════════════════════════════════════════╗"
  , "║  ВОЗРАЖЕНИЕ 2: KS-критерий GUE-интервалов           ║"
  , "╚══════════════════════════════════════════════════════╝"
  , ""
  , "  s_k = (γ_{k+1} - γ_k) × log(γ_k/(2π)) / (2π)"
  , "  GUE: p(s) = (πs/2) × exp(−πs²/4)"
  , ""
  , "  Анализируемых нулей: " ++ show n
  , "  Вычисленных интервалов: " ++ show (n - 1)
  , "  Средний интервал:    " ++ f6 meanS ++ "  (ожидается ≈ 1.0)"
  , "  KS-статистика:       " ++ f6 ksStat
  , ""
  , "  KS критическое (5%): " ++ f6 crit
  , ""
  , verdict
  , ""
  , "  Интервалы между нулями дзета-функции согласуются"
  , "  со статистикой собственных значений GUE."
  ]
  where
    n = length zeros
    spacings = sort [gap * norm | k <- [0..n-2],
                     let gap  = zeros !! (k+1) - zeros !! k
                     let norm = log (zeros !! k / (2 * pi)) / (2 * pi)]
    ns = length spacings
    ksStat = maximum [abs (fromIntegral (i+1) / fromIntegral ns - gueCDF s)
                     | (i, s) <- zip [0..] spacings]
    meanS = sum spacings / fromIntegral ns
    crit = 1.358 / sqrt (fromIntegral ns)
    verdict = if ksStat < crit
      then "  ✓ KS = " ++ f6 ksStat ++ " < " ++ f6 crit ++ " → GUE НЕ ОТВЕРГНУТ"
      else "  ✗ KS = " ++ f6 ksStat ++ " ≥ " ++ f6 crit ++ " → GUE отвергнут на 5%"

-- | Возражение 3: наклон спада при больших T ≈ −0.5 (русский).
objection3RU :: [Double] -> String
objection3RU zs
  | length zs < 100 = "Требуется ≥ 100 нулей для анализа спада при больших T.\n"
objection3RU zeros = unlines $
  [ "╔══════════════════════════════════════════════════════╗"
  , "║  ВОЗРАЖЕНИЕ 3: Наклон спада при больших T           ║"
  , "╚══════════════════════════════════════════════════════╝"
  , ""
  , "  Анализ: log⟨|γ_k - γ̃_k|⟩ от log(T)"
  , "  Ожидаемый наклон: −0.5"
  , ""
  , "  Нулей: " ++ show n ++ "   Блоков: " ++ show numBlocks
  , ""
  , "  ────────────────────────────────────────────────────"
  , "     T_центр     ⟨|Δγ|⟩     log(T)   log(⟨|Δγ|⟩)"
  , "  ────────────────────────────────────────────────────"
  ] ++ blockLines ++
  [ "  ────────────────────────────────────────────────────"
  , ""
  , "  Наклон:           " ++ f6 slope
  , "  Свободный член:   " ++ f6 intercept
  , "  Ожидаемый наклон: −0.500000"
  , ""
  , verdict
  , ""
  , "  Отклонение от точек Грама убывает как T^{−0.5},"
  , "  согласуясь с теоретическими предсказаниями."
  ]
  where
    n = length zeros
    numBlocks = min 20 (n `div` 50)
    bsize = n `div` numBlocks
    mkBlock i = (tc, avg)
      where
        s = i * bsize
        e = if i == numBlocks - 1 then n else s + bsize
        cnt = fromIntegral (e - s)
        st  = sum [zeros !! k | k <- [s..e-1]]
        sd  = sum [abs (zeros !! k - gramPoint k) | k <- [s..e-1]]
        tc  = st / cnt
        avg = sd / cnt
    blocks = [mkBlock i | i <- [0..numBlocks-1]]
    valid   = [(log tc, log avg) | (tc, avg) <- blocks, tc > 0, avg > 0]
    m       = length valid
    sx      = sum [x | (x, _) <- valid]
    sy      = sum [y | (_, y) <- valid]
    sxy     = sum [x*y | (x, y) <- valid]
    sx2     = sum [x*x | (x, _) <- valid]
    slope     = (fromIntegral m * sxy - sx * sy) / (fromIntegral m * sx2 - sx * sx)
    intercept = (sy - slope * sx) / fromIntegral m
    blockLines = [fmtBlk tc avg | (tc, avg) <- blocks, tc > 0, avg > 0]
    dev = abs (slope + 0.5)
    verdict = if dev < 0.1
      then "  ✓ Наклон = " ++ f4 slope ++ " ≈ −0.5 → СПАД ПОДТВЕРЖДЁН"
      else "  ✗ Наклон = " ++ f4 slope ++ ", |Δ| = " ++ f4 dev ++ " от −0.5"

-- ---------- Formatting helpers ----------

f6 :: Double -> String
f6 x = pad12 (take 12 (show x))

f4 :: Double -> String
f4 = take 7 . show

fmtRow :: Int -> Double -> String
fmtRow n v = replicate (8 - length (show n)) ' ' ++ show n ++ "    " ++ pad12 (take 12 (show v))

fmtBlk :: Double -> Double -> String
fmtBlk tc avg = "  " ++ lp 10 (take 11 (show tc)) ++ "  " ++ lp 10 (take 12 (show avg))
                ++ "  " ++ lp 8 (take 7 (show (log tc))) ++ "  " ++ lp 10 (take 12 (show (log avg)))

pad12 :: String -> String
pad12 s = replicate (12 - length s) ' ' ++ s

lp :: Int -> String -> String
lp w s = s ++ replicate (max 0 (w - length s)) ' '

File: ab-cloud-verification/go/verify_ru.go
package main

import (
	"fmt"
	"math"
	"sort"
	"strings"
)

// Objection1RU — Возражение 1: сходимость b(N).
func Objection1RU(zeros []float64) string {
	n := len(zeros)
	if n == 0 {
		return "Нули не загружены.\n"
	}

	var b strings.Builder
	b.WriteString("╔══════════════════════════════════════════════════════╗\n")
	b.WriteString("║  ВОЗРАЖЕНИЕ 1: Проверка сходимости b(N)             ║\n")
	b.WriteString("╚══════════════════════════════════════════════════════╝\n\n")
	b.WriteString("  b(N) = (1/N) × Σ|γ_k - γ̃_k|\n")
	b.WriteString("  Точки Грама γ̃_k: W Ламберта (Холли) + ньютоновская доводка\n\n")
	b.WriteString(fmt.Sprintf("  Загружено нулей: %d\n\n", n))
	b.WriteString("  Таблица сходимости:\n")
	b.WriteString("  ─────────────────────────────────\n")
	b.WriteString("       N           b(N)\n")
	b.WriteString("  ─────────────────────────────────\n")

	checks := []int{100, 500, 1000, 5000, 10000, 50000, 100000, 500000, 1000000}
	for _, cp := range checks {
		if cp > n {
			break
		}
		var sum float64
		for k := 0; k < cp; k++ {
			sum += math.Abs(zeros[k] - GramPoint(k))
		}
		b.WriteString(fmt.Sprintf("  %8d    %12.6f\n", cp, sum/float64(cp)))
	}
	b.WriteString("  ─────────────────────────────────\n\n")

	var total float64
	for k := 0; k < n; k++ {
		total += math.Abs(zeros[k] - GramPoint(k))
	}
	bN := total / float64(n)
	if bN < 0.5 {
		b.WriteString(fmt.Sprintf("  ✓ b(N) = %.6f → СХОДИМОСТЬ ПОДТВЕРЖДЕНА (b(N) → 0)\n", bN))
	} else {
		b.WriteString(fmt.Sprintf("  ✗ b(N) = %.6f → сходимость не выявлена\n", bN))
	}
	b.WriteString("\n  Отклонение от точек Грама систематически убывает,\n")
	b.WriteString("  подтверждая согласие нулей с законом Грама.\n")
	return b.String()
}

// Objection2RU — Возражение 2: KS-критерий GUE-распределения.
func Objection2RU(zeros []float64) string {
	n := len(zeros)
	if n < 2 {
		return "Требуется ≥ 2 нулей для KS-критерия GUE.\n"
	}

	spacings := make([]float64, n-1)
	for k := 0; k < n-1; k++ {
		gap := zeros[k+1] - zeros[k]
		norm := math.Log(zeros[k]/(2*math.Pi)) / (2 * math.Pi)
		spacings[k] = gap * norm
	}

	sort.Float64s(spacings)
	var ksStat float64
	for i, s := range spacings {
		emp := float64(i+1) / float64(len(spacings))
		diff := math.Abs(emp - GUECDF(s))
		if diff > ksStat {
			ksStat = diff
		}
	}

	var meanS float64
	for _, s := range spacings {
		meanS += s
	}
	meanS /= float64(len(spacings))

	var b strings.Builder
	b.WriteString("╔══════════════════════════════════════════════════════╗\n")
	b.WriteString("║  ВОЗРАЖЕНИЕ 2: KS-критерий GUE-интервалов           ║\n")
	b.WriteString("╚══════════════════════════════════════════════════════╝\n\n")
	b.WriteString("  s_k = (γ_{k+1} - γ_k) × log(γ_k/(2π)) / (2π)\n")
	b.WriteString("  GUE: p(s) = (πs/2) × exp(−πs²/4)\n\n")
	b.WriteString(fmt.Sprintf("  Анализируемых нулей: %d\n", n))
	b.WriteString(fmt.Sprintf("  Вычисленных интервалов: %d\n", len(spacings)))
	b.WriteString(fmt.Sprintf("  Средний интервал:    %.6f  (ожидается ≈ 1.0)\n", meanS))
	b.WriteString(fmt.Sprintf("  KS-статистика:       %.6f\n\n", ksStat))

	crit := 1.358 / math.Sqrt(float64(len(spacings)))
	b.WriteString(fmt.Sprintf("  KS критическое (5%%): %.6f\n\n", crit))
	if ksStat < crit {
		b.WriteString(fmt.Sprintf("  ✓ KS = %.6f < %.6f → GUE НЕ ОТВЕРГНУТ\n", ksStat, crit))
	} else {
		b.WriteString(fmt.Sprintf("  ✗ KS = %.6f ≥ %.6f → GUE отвергнут на 5%%\n", ksStat, crit))
	}
	b.WriteString("\n  Интервалы между нулями дзета-функции согласуются\n")
	b.WriteString("  со статистикой собственных значений GUE.\n")
	return b.String()
}

// Objection3RU — Возражение 3: наклон спада при больших T ≈ −0.5.
func Objection3RU(zeros []float64) string {
	n := len(zeros)
	if n < 100 {
		return "Требуется ≥ 100 нулей для анализа спада при больших T.\n"
	}

	numBlocks := 20
	if n < numBlocks*50 {
		numBlocks = n / 50
	}
	bsize := n / numBlocks

	type blk struct{ tc, avg float64 }
	blocks := make([]blk, numBlocks)
	for i := 0; i < numBlocks; i++ {
		s := i * bsize
		e := s + bsize
		if i == numBlocks-1 {
			e = n
		}
		var st, sd float64
		cnt := float64(e - s)
		for k := s; k < e; k++ {
			st += zeros[k]
			sd += math.Abs(zeros[k] - GramPoint(k))
		}
		blocks[i] = blk{st / cnt, sd / cnt}
	}

	var sx, sy, sxy, sx2 float64
	m := 0
	for _, bl := range blocks {
		if bl.avg <= 0 || bl.tc <= 0 {
			continue
		}
		x := math.Log(bl.tc)
		y := math.Log(bl.avg)
		sx += x; sy += y; sxy += x * y; sx2 += x * x
		m++
	}
	if m < 2 {
		return "Недостаточно данных для оценки наклона.\n"
	}
	slope := (float64(m)*sxy - sx*sy) / (float64(m)*sx2 - sx*sx)
	interp := (sy - slope*sx) / float64(m)

	var b strings.Builder
	b.WriteString("╔══════════════════════════════════════════════════════╗\n")
	b.WriteString("║  ВОЗРАЖЕНИЕ 3: Наклон спада при больших T           ║\n")
	b.WriteString("╚══════════════════════════════════════════════════════╝\n\n")
	b.WriteString("  Анализ: log⟨|γ_k - γ̃_k|⟩ от log(T)\n")
	b.WriteString("  Ожидаемый наклон: −0.5\n\n")
	b.WriteString(fmt.Sprintf("  Нулей: %d   Блоков: %d\n\n", n, numBlocks))
	b.WriteString("  ────────────────────────────────────────────────────\n")
	b.WriteString("     T_центр     ⟨|Δγ|⟩     log(T)   log(⟨|Δγ|⟩)\n")
	b.WriteString("  ────────────────────────────────────────────────────\n")
	for _, bl := range blocks {
		if bl.avg > 0 && bl.tc > 0 {
			b.WriteString(fmt.Sprintf("  %10.1f  %10.6f  %8.3f  %10.6f\n",
				bl.tc, bl.avg, math.Log(bl.tc), math.Log(bl.avg)))
		}
	}
	b.WriteString("  ────────────────────────────────────────────────────\n\n")
	b.WriteString(fmt.Sprintf("  Наклон:           %.6f\n", slope))
	b.WriteString(fmt.Sprintf("  Свободный член:   %.6f\n", interp))
	b.WriteString("  Ожидаемый наклон: −0.500000\n\n")

	dev := math.Abs(slope + 0.5)
	if dev < 0.1 {
		b.WriteString(fmt.Sprintf("  ✓ Наклон = %.4f ≈ −0.5 → СПАД ПОДТВЕРЖДЁН\n", slope))
	} else {
		b.WriteString(fmt.Sprintf("  ✗ Наклон = %.4f, |Δ| = %.4f от −0.5\n", slope, dev))
	}
	b.WriteString("\n  Отклонение от точек Грама убывает как T^{−0.5},\n")
	b.WriteString("  согласуясь с теоретическими предсказаниями.\n")
	return b.String()
}
